### PR TITLE
Otel server context interceptor

### PR DESCRIPTION
**Pull request by YifeiZhuang**
b/357892637
following up on: https://github.com/grpc/grpc-java/pull/11477
This uses inprocess transport. Next we will add interop test using netty transport to test real across thread span propagation.

### Context
`io.opentelemetry.context` forks `io.grpc.context`, so propagating span context becomes an issue.
In grpc, the server context is propagated from transport thread to the application thread at the `JumpToApplicationThreadServerListener`, using `ContextRunnable`. However, due to opentelemetry has its own context object/storage, the span propagation does not happen automatically with the existing grpc support. There are two ways of doing this:
1. (this PR) use the interceptor to explicitly propagate span to the application thread. (interceptor runs in application thread, `serverStreamTracer.filterContext` runs in transport thread).  
*Pros*: simplicity, UX, API support 
*Cons*:context does not propagate between opentelemetry and grpc, only span context is manually propagated.
4. Implement `io.grpc.override.ContextStorageOverride`, e.g. https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/grpc-1.6/library/src/main/java/io/grpc/override/ContextStorageOverride.java,
 which delegates the grpc context storage to opentelemetry context.
*Pros*: propagate all 
*Cons*: UX is complicated. users can not it use with other custom ContextStorageOverride if we add it to grpc-opentelemetry package. Or if we release it as a separate artifact, users need to add dependency, no API support.

Client side is easier, the current span context in opentelemetry from the application will be propagated as parent span towards grpc client/attempt spans, automatically support by SpanBuilderImpl.

### CODE REVIEW

**Comment by YifeiZhuang** [comment]
In the interop-test, I switched to `GrpcTraceBinPropagator`, but then I have to disable census in order to verify the trace-id is propagated from client application span -> grpc client span -> grpc attempt span -> grpc server span -> server application span. However, there seems no API we can disable census if we keep it in the dependency?

File: opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java [hunk 1]
```diff
@@ -315,9 +316,59 @@ final class TracingServerSpanPropagationInterceptor implements ServerInterceptor
     public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(ServerCall<ReqT, RespT> call,
         Metadata headers, ServerCallHandler<ReqT, RespT> next) {
       Span span = otelSpan.get(io.grpc.Context.current());
-      try (Scope scope = Context.current().with(span).makeCurrent()) {
+      if (span == null) {
+        logger.log(Level.FINE, "Server span not found. ServerTracerFactory for server "
+            + "tracing must be set.");
         return next.startCall(call, headers);
       }
+      try (Scope scope = Context.current().with(span).makeCurrent()) {
```
Comment (ejona86) [inline opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java:324]: Do the `Context.current().with(span)` once in this method, then call `makeCurrent()` on that single instance in all the places. The context shouldn't be changing between the callbacks.

File: opentelemetry/src/main/java/io/grpc/opentelemetry/GrpcOpenTelemetry.java [hunk 2]
```diff
@@ -152,6 +154,7 @@ public void configureChannelBuilder(ManagedChannelBuilder<?> builder) {
     InternalManagedChannelBuilder.addMetricSink(builder, sink);
     InternalManagedChannelBuilder.interceptWithTarget(
         builder, openTelemetryMetricsModule::getClientInterceptor);
+    builder.intercept(openTelemetryTracingModule.getClientInterceptor());
```
Comment (ejona86) [inline opentelemetry/src/main/java/io/grpc/opentelemetry/GrpcOpenTelemetry.java:157]: We should avoid enabling tracing until we are happy with it. Are we happy with it right after this PR?

You could add a package-private builder option to allow enabling it for testing and such.

Comment (YifeiZhuang) [reply opentelemetry/src/main/java/io/grpc/opentelemetry/GrpcOpenTelemetry.java:157]: No I don't think we should make the API public and stable right away.

Let me split the PR and have interop test in a separate PR.
The interop test needs it to be public API though, how about we have environmental variable and disable by default ?

Comment (ejona86) [reply opentelemetry/src/main/java/io/grpc/opentelemetry/GrpcOpenTelemetry.java:157]: Env var is fine, but you can't change env vars from Java. You could use system properties instead.

But I'd probably just make the package-private method on the builder and then make an InternalGrpcOpenTelemetry class that exposes the method for the test to use.

**Review by ejona86 — APPROVED** [review]

**Pull request merged**